## We covered a lot of information today and I'd like you to practice developing classification trees on your own. For each exercise, work through the problem, determine the result, and provide the requested interpretation in comments along with the code. The point is to build classifiers, not necessarily good classifiers (that will hopefully come later)

In [189]:
import numpy as np
import pandas as pd
import pydotplus
import matplotlib.pyplot as plt
%matplotlib inline

from pandas.tools.plotting import scatter_matrix
from sklearn import datasets
from sklearn import tree
from sklearn.cross_validation import train_test_split
from sklearn import metrics
from sklearn.externals.six import StringIO

### 1. Load the iris dataset and create a holdout set that is 50% of the data (50% in training and 50% in test). Output the results (don't worry about creating the tree visual unless you'd like to) and discuss them briefly (are they good or not?)

In [190]:
iris = datasets.load_iris()
iris

{'DESCR': 'Iris Plants Database\n\nNotes\n-----\nData Set Characteristics:\n    :Number of Instances: 150 (50 in each of three classes)\n    :Number of Attributes: 4 numeric, predictive attributes and the class\n    :Attribute Information:\n        - sepal length in cm\n        - sepal width in cm\n        - petal length in cm\n        - petal width in cm\n        - class:\n                - Iris-Setosa\n                - Iris-Versicolour\n                - Iris-Virginica\n    :Summary Statistics:\n\n    ============== ==== ==== ======= ===== ====================\n                    Min  Max   Mean    SD   Class Correlation\n    ============== ==== ==== ======= ===== ====================\n    sepal length:   4.3  7.9   5.84   0.83    0.7826\n    sepal width:    2.0  4.4   3.05   0.43   -0.4194\n    petal length:   1.0  6.9   3.76   1.76    0.9490  (high!)\n    petal width:    0.1  2.5   1.20  0.76     0.9565  (high!)\n    ============== ==== ==== ======= ===== ====================\n\n

In [193]:
x = iris.data[:,2:]
y = iris.target
dt = tree.DecisionTreeClassifier()

In [194]:
x_train, x_test, y_train, y_test = train_test_split(x,y,test_size=0.5,train_size=0.5)

In [195]:
dt = dt.fit(x_train,y_train)

In [196]:
def measure_performance(x,y,dt, show_accuracy=True, show_classification_report=True, show_confusion_matrix=True):
    y_pred=dt.predict(x)
    if show_accuracy:
        print("Accuracy:{0:.3f}".format(metrics.accuracy_score(y, y_pred)),"\n")
    if show_classification_report:
        print("Classification report")
        print(metrics.classification_report(y,y_pred),"\n")
    if show_confusion_matrix:
        print("Confusion matrix")
        print(metrics.confusion_matrix(y,y_pred),"\n")

In [197]:
measure_performance(x_test,y_test,dt)

Accuracy:0.947 

Classification report
             precision    recall  f1-score   support

          0       1.00      1.00      1.00        30
          1       0.93      0.93      0.93        28
          2       0.88      0.88      0.88        17

avg / total       0.95      0.95      0.95        75
 

Confusion matrix
[[30  0  0]
 [ 0 26  2]
 [ 0  2 15]] 



Nearly 95%
accuracy, with 100% precision for the first species (0), and progressively less precision for the latter two species (1, 2) . 

30 were classified as species 1, 26 as species 2, and 15 as species 3; though there were two falsely classified samples in both species 2 and species 3. 

### 2. Redo the model with a 75% - 25% training/test split and compare the results. Are they better or worse than before? Discuss why this may be.

In [198]:
x_train, x_test, y_train, y_test = train_test_split(x,y,test_size=0.25,train_size=0.75)
dt = dt.fit(x_train,y_train)
y_pred=dt.predict(x_test)
measure_performance(x_test,y_test,dt)

Accuracy:0.974 

Classification report
             precision    recall  f1-score   support

          0       1.00      1.00      1.00        12
          1       0.93      1.00      0.97        14
          2       1.00      0.92      0.96        12

avg / total       0.98      0.97      0.97        38
 

Confusion matrix
[[12  0  0]
 [ 0 14  0]
 [ 0  1 11]] 



Over 97% accuracy, with again 100% precision for the first species (0), and a similar percent accuracy for the latter two species (1, 2) . 

12 were classified as species 1, 14 as species 2, and 11 as species 3; and there was one falsely classified sample in species 3. 

There are less total samples because the testing size is larger.

### 3. Load the breast cancer dataset (`datasets.load_breast_cancer()`) and perform basic exploratory analysis. What attributes to we have? What are we trying to predict?
For context of the data, see the documentation here: https://archive.ics.uci.edu/ml/datasets/Breast+Cancer+Wisconsin+%28Diagnostic%29

In [199]:
bc = datasets.load_breast_cancer()
bc

{'DESCR': 'Breast Cancer Wisconsin (Diagnostic) Database\n\nNotes\n-----\nData Set Characteristics:\n    :Number of Instances: 569\n\n    :Number of Attributes: 30 numeric, predictive attributes and the class\n\n    :Attribute Information:\n        - radius (mean of distances from center to points on the perimeter)\n        - texture (standard deviation of gray-scale values)\n        - perimeter\n        - area\n        - smoothness (local variation in radius lengths)\n        - compactness (perimeter^2 / area - 1.0)\n        - concavity (severity of concave portions of the contour)\n        - concave points (number of concave portions of the contour)\n        - symmetry \n        - fractal dimension ("coastline approximation" - 1)\n        \n        The mean, standard error, and "worst" or largest (mean of the three\n        largest values) of these features were computed for each image,\n        resulting in 30 features.  For instance, field 3 is Mean Radius, field\n        13 is Rad

The attributes (x) are listed near the top of the dataset, and include radius, texture, perimeter, etc. 
The target (y) is what we are trying to predict. Here, that is whether a tumor is malignant or benign. 

In [200]:
x = bc.data[:,:] 
y = bc.target 
x_train, x_test, y_train, y_test = train_test_split(x,y,test_size=0.5,train_size=0.5)
dt = dt.fit(x_train,y_train)

In [201]:
def measure_performance(x,y,dt, show_accuracy=True, show_classification_report=True, show_confusion_matrix=True):
    y_pred=dt.predict(x)
    if show_accuracy:
        print("Accuracy:{0:.5f}".format(metrics.accuracy_score(y, y_pred)),"\n")
    if show_classification_report:
        print("Classification report")
        print(metrics.classification_report(y,y_pred),"\n")
    if show_confusion_matrix:
        print("Confusion matrix")
        print(metrics.confusion_matrix(y,y_pred),"\n")

In [202]:
measure_performance(x_test,y_test,dt)

Accuracy:0.92632 

Classification report
             precision    recall  f1-score   support

          0       0.98      0.83      0.90       112
          1       0.90      0.99      0.94       173

avg / total       0.93      0.93      0.93       285
 

Confusion matrix
[[ 93  19]
 [  2 171]] 



### 4. Using the breast cancer data, create a classifier to predict the type of seed. Perform the above hold out evaluation (50-50 and 75-25) and discuss the results.

In [203]:
x_train, x_test, y_train, y_test = train_test_split(x,y,test_size=0.25,train_size=0.75)
dt = dt.fit(x_train,y_train)

In [204]:
measure_performance(x_test,y_test,dt)

Accuracy:0.95105 

Classification report
             precision    recall  f1-score   support

          0       0.98      0.89      0.93        53
          1       0.94      0.99      0.96        90

avg / total       0.95      0.95      0.95       143
 

Confusion matrix
[[47  6]
 [ 1 89]] 



Over 95% accuracy, with rougly 95% precision. 7 out of 143 samples were misclassified. 